### PR TITLE
fix(s3): add missing safe actions to `grantWrite`, `grantReadWrite` and `grantPut` methods

### LINKED ISSUE #13616: aws-s3: grantPut not working as advertised
The `grantPut` method documentation reads:

> Grants s3:PutObject* and s3:Abort* permissions for this bucket to an IAM principal.
> 
> If encryption is used, permission to use the key to encrypt the contents of written files will also be granted to the same principal.

### Reproduction Steps

```typescript
const function = new lambda.Function(this, "FunkyTown", {...});
const bucket = new s3.Bucket(this, "AWholeLotOfBucket");
bucket.grantPut(function);
```

### What did you expect to happen?

To add `action` `s3:PutObject*` thus:

```yaml
Action:
  - s3:PutObject*
  - s3:Abort*
Effect: Allow
Resource:
  Fn::Join:
    - ""
    - - Fn::GetAtt:
          - AWholeLotOfBucketA765543B
          - Arn
      - /*
```

### What actually happened?

```yaml
Action:
  - s3:PutObject
  - s3:Abort*
Effect: Allow
Resource:
  Fn::Join:
    - ""
    - - Fn::GetAtt:
          - AWholeLotOfBucketA765543B
          - Arn
      - /*
```

### Environment

  - **CDK CLI Version  : 1.93.0
  - **Framework Version: 1.93.0
  - **Node.js Version: v14.16.0 
  - **OS               : Amazon Linux release 2 (Karoo)
  - **Language (Version): 3.9.9 

### Other

Link to docs: https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.Bucket.html#grantwbrputidentity-objectskeypattern
Link to original PR: https://github.com/aws/aws-cdk/pull/591

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by flavioleggio**
In this pull request I try to fix a big behavioral change in the bucket grant methods for write actions. This change was introduced due to security issues as stated in [this pull request](https://github.com/aws/aws-cdk/pull/12391) and substituted the `s3:PutObject*` action glob pattern with the simple `s3:PutObject` to exclude the dangerous `s3:PutObjectAcl` and `s3:PutObjectVersionAcl` from the equation.

While I agree about the security issues, I think that other useful actions - if safe, obviously - should be added when calling the `grantWrite`, `grantReadWrite` and `grantPut` methods. This pull request adds the following actions:
* s3:PutObjectLegalHold
* s3:PutObjectRetention
* s3:PutObjectTagging
* s3:PutObjectVersionTagging

I also added the `s3:PutObjectVersionAcl` action to the `grantPutAcl` method, along with the existing `s3:PutObjectAcl`.

I adapted existing unit and integ tests to accept these new actions.

Fixes #13616

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by otaviomacedo** [comment]
Thanks, @flavioleggio. We got approval from AppSec for this. Can you just resolve the conflicts please?

**Comment by flavioleggio** [comment]
@otaviomacedo done!

File: packages/decdk/bin/decdk-schema.d.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export {};
```
Comment (otaviomacedo) [inline packages/decdk/bin/decdk-schema.d.ts:1]: These decdk files should not be here.

**Review by otaviomacedo — CHANGES_REQUESTED** [review]

File: packages/decdk/bin/decdk-schema.d.ts (hunk 1 above)
Comment (flavioleggio) [reply packages/decdk/bin/decdk-schema.d.ts:1]: Maybe this is fault of conflicts from a previous merge. Should I delete the file?

Comment (flavioleggio) [reply packages/decdk/bin/decdk-schema.d.ts:1]: Ok, I see what you mean. All decdk files should be rolled back to master state. I really don't know why these files ended up here, I did not manually edited them. I will fix this immediately.

Comment (otaviomacedo) [reply packages/decdk/bin/decdk-schema.d.ts:1]: Yes, please. `decdk` has been moved to a different repository.

Comment (flavioleggio) [reply packages/decdk/bin/decdk-schema.d.ts]: Done.

**Review by otaviomacedo — APPROVED** [review]

**Comment by quixoticmonk** [comment]
Shouldn't the api docs reflect this change ? https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3.Bucket.html#grantwbrputidentity-objectskeypattern
States `Grants s3:PutObject* and s3:Abort* permissions for this bucket to an IAM principal.`

**Pull request merged**